def test_2_databases():
    from . import testing_functions
    import logging
    import ontoweaver

    logger = logging.getLogger("ontoweaver")
    logger.setLevel(logging.DEBUG)

    directory_name1 = "oncokb"
    directory_name2 = "multi_type_transformer"

    assert_nodes = [('0:variant', 'variant', {'whatever': 'A0'}),
('A:disease', 'disease', {'whatever': 'A0', 'something': 'Whatever it is'}),
('1:variant', 'variant', {'whatever': 'B1'}),
('B:patient', 'patient', {'something': 'Whatever it is'}),
('2:variant', 'variant', {'whatever': 'C2'}),
('C:oncogenicity', 'oncogenicity', {}),
('0:variant', 'variant', {'timestamp': '2023-02-27 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'MET amplification results from the gain of the MET gene on chromosome 7q31 and is found in various cancers, including melanoma, lung and renal cancers (PMID: 28481359). Often, this amplification leads to overexpression of the MET protein, which has been demonstrated to contribute to tumorigenesis by increasing downstream pathway activation, proliferation, invasiveness, and angiogenesis (PMID: 16461907, 18077425, 17463250, 19117057, 22869872,17667909). MET amplification has also been shown to contribute to resistance to tyrosine kinase inhibitors, including gefitinib and erlotinib and more recently osimertinib, as shown by sustained pathway activation and proliferation in the presence of drug (PMID: 18093943, 17463250, 30073261). In vitro proliferation assays of cell lines with MET amplifications show sensitivity to the MET-specific kinase inhibitor, PHA-665752 (PMID: 17463250, 16461907).', 'variant_summary': 'MET amplification is known to be oncogenic.'}),
('33:patient', 'patient', {}),
('MET:gene_hugo', 'gene_hugo', {'entrez_gene_id': '4233', 'gene_summary': 'MET, a receptor tyrosine kinase, is recurrently altered by mutation, amplification and rarely altered by gene translocation in various cancer types.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'While crizotinib is NCCN-compendium listed for the treatment of patients with high-level MET-amplified non-small cell lung cancer, its clinical utility in patients with MET-amplified high-grade serous ovarian cancer is unknown.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('30073261:publication', 'publication', {}),
('16461907:publication', 'publication', {}),
('19117057:publication', 'publication', {}),
('17667909:publication', 'publication', {}),
('22869872:publication', 'publication', {}),
('18077425:publication', 'publication', {}),
('18093943:publication', 'publication', {}),
('28481359:publication', 'publication', {}),
('17463250:publication', 'publication', {}),
('Crizotinib:drug', 'drug', {}),
('Telisotuzumab Vedotin:drug', 'drug', {}),
('Capmatinib:drug', 'drug', {}),
('Tepotinib:drug', 'drug', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('91:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('33:patient', 'patient', {}),
('LEVEL_Fda3:fda_evidence_level', 'fda_evidence_level', {}),
('LEVEL_3B:oncokb_evidence_level', 'oncokb_evidence_level', {}),
('1:variant', 'variant', {'timestamp': '2023-06-22 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'BRAF amplification is a rare event in cancer and has been found to occur at very low frequency in melanoma and glioma (cBioPortal, MSKCC, April 2017). It has been demonstrated that amplification of wildtype BRAF in melanoma cells leads to overexpression of the protein and constitutive activation of MAPK pathway, demonstrated as a BRAF-dependent increase in ERK phosphorylation, as well as an increase in cell proliferation (PMID: 15467732). BRAF overexpression has been associated with increased proliferation in gliomas (PMID: 25618114). Amplification of BRAF V600E allele has been described as a mechanism of resistance to MEK/RAF inhibitors in melanoma via MAPK pathway hyperactivation (PMID: 22395615, 24055054, 21098728).', 'variant_summary': 'BRAF amplification is known to be oncogenic.'}),
('67:patient', 'patient', {}),
('BRAF:gene_hugo', 'gene_hugo', {'entrez_gene_id': '673', 'gene_summary': 'BRAF, an intracellular kinase, is frequently mutated in melanoma, thyroid and lung cancers among others.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with BRAF-amplified high-grade serous ovarian cancer.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('24055054:publication', 'publication', {}),
('25618114:publication', 'publication', {}),
('22395615:publication', 'publication', {}),
('21098728:publication', 'publication', {}),
('15467732:publication', 'publication', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('4:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('67:patient', 'patient', {}),
('2:variant', 'variant', {'timestamp': '2020-11-16 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'EZH2 amplification results in overexpression of the protein. This alteration has been found in several tumor types and has been implicated as a marker of poor prognosis in various solid tumors (PMID: 12374981, 14500907, 21837672, 16330673, 16330673). EZH2 overexpression in natural killer and natural killer T-cell lymphoma cell lines resulted in a growth advantage compared to wildtype (PMID: 23529930).', 'variant_summary': 'EZH2 amplification is known to be oncogenic.'}),
('88:patient', 'patient', {}),
('EZH2:gene_hugo', 'gene_hugo', {'entrez_gene_id': '2146', 'gene_summary': 'EZH2, an epigenetic modifier, is altered by mutation and/or overexpression in solid tumors and hematologic malignancies.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with EZH2-amplified high-grade serous ovarian cancer.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('16330673:publication', 'publication', {}),
('12374981:publication', 'publication', {}),
('23529930:publication', 'publication', {}),
('14500907:publication', 'publication', {}),
('21837672:publication', 'publication', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('92:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('88:patient', 'patient', {}),
('3:variant', 'variant', {'timestamp': '2023-02-15 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'Deletion of CDKN2A results in loss of the p16/INK4A and p14/ARF proteins. CDKN2A deletion has been found in head and neck squamous cell carcinoma, glioma, bladder carcinoma, leukemia and melanoma and is often associated with poorer clinical outcomes (PMID: 15495191, 26516359, 10999737). Deletion of CDKN2A in mouse models of malignant mesothelioma resulted in accelerated tumorigenesis compared to wildtype (PMID: 21526190). In addition, knockdown of CDKN2A in an HRAS G12V mouse model resulted in the induction of high-grade endometrial stromal sarcoma (PMID: 28982163).', 'variant_summary': 'CDKN2A deletion is known to be oncogenic.'}),
('60:patient', 'patient', {}),
('CDKN2A:gene_hugo', 'gene_hugo', {'entrez_gene_id': '1029', 'gene_summary': 'The CDKN2A gene encodes two proteins, p16INK4A and p14ARF, that regulate the cell growth and survival. CDKN2A is altered by mutation and/or deletion in a broad range of solid and hematologic cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'Laboratory data suggest that cancer cells with loss-of-function alterations of CDKN2A may be sensitive to CDK4/6 inhibitors such as palbociclib, ribociclib and abemaciclib.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('21526190:publication', 'publication', {}),
('28982163:publication', 'publication', {}),
('26516359:publication', 'publication', {}),
('15495191:publication', 'publication', {}),
('10999737:publication', 'publication', {}),
('Palbociclib:drug', 'drug', {}),
('Ribociclib:drug', 'drug', {}),
('Abemaciclib:drug', 'drug', {}),
('Loss-of-function:functional_effect', 'functional_effect', {}),
('37:sample', 'sample', {}),
('Deletion:alteration', 'alteration', {}),
('60:patient', 'patient', {}),
('LEVEL_Fda3:fda_evidence_level', 'fda_evidence_level', {}),
('LEVEL_4:oncokb_evidence_level', 'oncokb_evidence_level', {}),
('4:variant', 'variant', {'timestamp': '2019-02-06 00:00:00', 'version': 'v4.10', 'variant_summary': 'ETV6 amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('32:patient', 'patient', {}),
('ETV6:gene_hugo', 'gene_hugo', {'entrez_gene_id': '2120', 'gene_summary': 'ETV6, a transcription factor, is frequently altered by chromosomal rearrangement in hematologic malignancies.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with ETV6-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('15:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('32:patient', 'patient', {}),
('5:variant', 'variant', {'timestamp': '2018-08-01 00:00:00', 'version': 'v4.10', 'variant_summary': 'ETNK1 amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('40:patient', 'patient', {}),
('ETNK1:gene_hugo', 'gene_hugo', {'entrez_gene_id': '55500', 'gene_summary': 'ETNK1, a kinase involved in phospholipid biosynthesis, is recurrently altered by mutation in BCR-ABL negative CML and CMML.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with ETNK1-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('43:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('40:patient', 'patient', {}),
('6:variant', 'variant', {'timestamp': '2021-11-18 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'KRAS amplification results in overexpression of the protein. This mutation has been found in germ cell tumors and esophagogastric and ovarian cancer (cBioPortal, Zehir, A et al. Nature Medicine, 2017). In vitro studies have demonstrated that this mutation confers resistance to the tyrosine kinase inhibitor crizotinib, EGFR inhibitors cetuximab or panitumumab, MET inhibitors PHA-665752 or JNJ38877605 and MEK inhibitors AZD6244 or GSK1120212 in patient-derived lung cancer cell line and xenograft models, colorectal cancer cell line models, MET-amplified cell line models and gastric cancer cell line models (PMID: 29808010, 30072474, 23404247, 20841479). Analyses of 475 patients with endometrial cancer demonstrated that KRAS amplification is associated with metastasis (PMID: 23099803).', 'variant_summary': 'KRAS amplification is likely oncogenic.'}),
('68:patient', 'patient', {}),
('KRAS:gene_hugo', 'gene_hugo', {'entrez_gene_id': '3845', 'gene_summary': 'KRAS, a GTPase which functions as an upstream regulator of the MAPK pathway, is frequently mutated in various cancer types including lung, colorectal and pancreatic cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with KRAS-amplified high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('23099803:publication', 'publication', {}),
('29808010:publication', 'publication', {}),
('30072474:publication', 'publication', {}),
('20841479:publication', 'publication', {}),
('23404247:publication', 'publication', {}),
('Likely Gain-of-function:functional_effect', 'functional_effect', {}),
('76:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('68:patient', 'patient', {}),
('7:variant', 'variant', {'timestamp': '2021-02-17 00:00:00', 'version': 'v4.10', 'variant_summary': 'NTRK3 amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('65:patient', 'patient', {}),
('NTRK3:gene_hugo', 'gene_hugo', {'entrez_gene_id': '4916', 'gene_summary': 'NTRK3, a receptor tyrosine kinase, is altered by gene fusion in various cancer types.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with NTRK3-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('46:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('65:patient', 'patient', {}),
('8:variant', 'variant', {'timestamp': '2019-02-04 00:00:00', 'version': 'v4.10', 'variant_summary': 'IDH2 amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('35:patient', 'patient', {}),
('IDH2:gene_hugo', 'gene_hugo', {'entrez_gene_id': '3418', 'gene_summary': 'IDH2, a cell metabolism enzyme, is recurrently mutated in various cancer types including acute myeloid leukemia, glioblastoma, and cholangiocarcinoma.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with IDH2-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('13:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('35:patient', 'patient', {}),
('9:variant', 'variant', {'timestamp': '2019-01-30 00:00:00', 'version': 'v4.10', 'variant_summary': 'MAF amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('75:patient', 'patient', {}),
('MAF:gene_hugo', 'gene_hugo', {'entrez_gene_id': '4094', 'gene_summary': 'MAF, a transcription factor, is recurrently altered by rearrangement in multiple myeloma.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with MAF-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('85:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('75:patient', 'patient', {}),
('10:variant', 'variant', {'timestamp': '2023-07-05 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'The mutation effect description for truncating mutations in BRCA1 is: Truncating mutations in BRCA1 can lead to varying C-terminally truncated proteins that results in aberrant protein folding, contributing to loss of BRCA1 protein function. Human breast cancer cell lines that contain a truncating mutation in BRCA1 display elevated levels of aneuploidy and impaired DNA damage response. In addition, truncating mutations have been shown to induce aberrant protein localization, which may impact the interaction of important binding partners (PMID: 20608970). Mouse models of BRCA1 truncating mutations develop cancer, including mammary carcinomas, lymphomas and ovarian carcinomas (PMID: 11358863, 12483515, 12947386).', 'variant_summary': 'The BRCA1 Y1463* is a truncating mutation in a tumor suppressor gene, and therefore is likely oncogenic.'}),
('61:patient', 'patient', {}),
('BRCA1:gene_hugo', 'gene_hugo', {'entrez_gene_id': '672', 'gene_summary': 'BRCA1, a tumor suppressor involved in the DNA damage response, is mutated in various cancer types.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'The PARP inhibitors olaparib, rucaparib and niraparib are FDA-approved for the treatment of patients with ovarian, fallopian tube, and primary peritoneal cancer with deleterious germline or somatic BRCA1 or 2 mutations (olaparib and niraparib are also approved for suspected deleterious mutations in this setting).'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('11358863:publication', 'publication', {}),
('20608970:publication', 'publication', {}),
('12483515:publication', 'publication', {}),
('12947386:publication', 'publication', {}),
('Olaparib:drug', 'drug', {}),
('Rucaparib:drug', 'drug', {}),
('Niraparib:drug', 'drug', {}),
('Talazoparib:drug', 'drug', {}),
('Likely Loss-of-function:functional_effect', 'functional_effect', {}),
('8:sample', 'sample', {}),
('Y1463*:alteration', 'alteration', {'alteration_consequence': 'stop_gained', 'protein_start': '1463.0', 'protein_end': '1463.0'}),
('61:patient', 'patient', {}),
('LEVEL_Fda2:fda_evidence_level', 'fda_evidence_level', {}),
('LEVEL_1:oncokb_evidence_level', 'oncokb_evidence_level', {}),
('11:variant', 'variant', {'timestamp': '2023-01-24 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'The mutation effect description for truncating mutations in TP53 is: Truncating mutations of TP53 occur throughout the gene and lead to the production of several C-terminally truncated protein forms. These alterations are predicted to be inactivating and are associated with poor prognosis (PMID: 11900253, 11753428, 16007150, 21467160, 19336573). Experimental studies have revealed that truncating mutations promote cancer cell proliferation, survival and metastasis, since ectopic expression of these mutations in melanoma cells increased cell motility and tumor formation in vivo. This was due in part to aberrant localization of truncated proteins to the mitochondria, regulating genes involved in cell survival, including CypD (PMID: 27759562).', 'variant_summary': 'There is no available functional data about the TP53 X125_splice alteration (last reviewed on 01/24/2023). However, it has been identified as a statistically significant hotspot and is likely to be oncogenic.'}),
('11:patient', 'patient', {}),
('TP53:gene_hugo', 'gene_hugo', {'entrez_gene_id': '7157', 'gene_summary': 'TP53, a tumor suppressor in the DNA damage pathway, is the most frequently mutated gene in cancer.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with TP53 X125_splice altered high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('19336573:publication', 'publication', {}),
('27759562:publication', 'publication', {}),
('16007150:publication', 'publication', {}),
('11753428:publication', 'publication', {}),
('11900253:publication', 'publication', {}),
('21467160:publication', 'publication', {}),
('Likely Loss-of-function:functional_effect', 'functional_effect', {}),
('1:sample', 'sample', {}),
('X125_splice:alteration', 'alteration', {'alteration_consequence': 'splice_region_variant', 'protein_start': '125.0', 'protein_end': '125.0'}),
('11:patient', 'patient', {}),
('12:variant', 'variant', {'timestamp': '2019-01-30 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'The mutation effect description for truncating mutations in BCOR is: Truncating mutations in BCOR occur as both germline (oculofaciocardiodental or OFCD syndrome) and somatic (acute myeloid leukemia (AML) myelodysplastic syndrome, retinoblastoma) mutations in human disease, leading to reduced protein expression (PMID: 22012066, 24047651, 25550361, 22237022). Experimental studies using AML patient samples suggest loss of BCOR protein enhances cell proliferation and myeloid differentiation (PMID: 26847029).', 'variant_summary': 'The BCOR S1704Kfs*16 is a truncating mutation in a tumor suppressor gene, and therefore is likely oncogenic.'}),
('77:patient', 'patient', {}),
('BCOR:gene_hugo', 'gene_hugo', {'entrez_gene_id': '54880', 'gene_summary': 'BCOR, a transcriptional repressor, is altered in various solid and hematologic malignancies including acute myeloid leukemia.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with BCOR S1704Kfs*16 mutant high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('26847029:publication', 'publication', {}),
('22012066:publication', 'publication', {}),
('25550361:publication', 'publication', {}),
('22237022:publication', 'publication', {}),
('24047651:publication', 'publication', {}),
('Likely Loss-of-function:functional_effect', 'functional_effect', {}),
('52:sample', 'sample', {}),
('S1704Kfs*16:alteration', 'alteration', {'alteration_consequence': 'frameshift_variant', 'protein_start': '1703.0', 'protein_end': '1703.0'}),
('77:patient', 'patient', {}),
('13:variant', 'variant', {'timestamp': '2023-06-21 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'FGFR1 amplification results in overexpression of the protein. This mutation has been found in squamous cell lung, bladder and breast cancer (PMID: 22960745, 26451490, 24476821). In vitro studies demonstrate that FGFR1 overexpression in breast cancer cell lines promotes resistance to the ESR1 inhibitor fulvestrant as measured by sustained cell viability upon drug treatment (PMID: 31371343). A patient with FGFR1-amplified small cell lung cancer had a partial response to the FGFR-inhibitor pazopanib (Abstract: Russo et al. JCO PO, 2019. https://ascopubs.org/doi/full/10.1200/PO.19.00243).', 'variant_summary': 'FGFR1 amplification is known to be oncogenic.'}),
('5:patient', 'patient', {}),
('FGFR1:gene_hugo', 'gene_hugo', {'entrez_gene_id': '2260', 'gene_summary': 'FGFR1, a receptor tyrosine kinase, is altered by mutation, chromosomal rearrangement or amplification in various cancer types including lung and breast cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'While there is promising clinical data supporting the use of pan-FGFR-targeted inhibitors such as erdafitinib in patients with high-level FGFR1-amplified squamous cell lung cancer, their clinical utility in patients with FGFR1-amplified high-grade serous ovarian cancer is unknown.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('31371343:publication', 'publication', {}),
('22960745:publication', 'publication', {}),
('24476821:publication', 'publication', {}),
('26451490:publication', 'publication', {}),
('Erdafitinib:drug', 'drug', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('1:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('5:patient', 'patient', {}),
('LEVEL_Fda3:fda_evidence_level', 'fda_evidence_level', {}),
('LEVEL_3B:oncokb_evidence_level', 'oncokb_evidence_level', {}),
('14:variant', 'variant', {'timestamp': '2019-08-28 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'Amplification of MYC is correlated with an increase in the expression of the MYC gene (PMID: 22464321,15083194), which leads to the transcriptional activation of MYC target genes. MYC target genes regulate various cellular processes involved in oncogenesis including proliferation, inhibiting exit from the cell cycle, stimulating vascularization and enhancing genomic instability (PMID: 22464321,10378696,16934487,19029958). Genomic amplification of MYC is observed in a large number of tumor types. A study of genomic amplification across many cancer types identified focal amplifications of MYC in 14% of all cancers, with additional enrichment for large scale arm-level copy number alterations encompassing the MYC gene locus (PMID: 20164920).', 'variant_summary': 'MYC amplification is known to be oncogenic.'}),
('94:patient', 'patient', {}),
('MYC:gene_hugo', 'gene_hugo', {'entrez_gene_id': '4609', 'gene_summary': 'MYC, a transcription factor, is altered by chromosomal rearrangement, amplification and overexpression in a variety of cancer types.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with MYC-amplified high-grade serous ovarian cancer.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('20164920:publication', 'publication', {}),
('16934487:publication', 'publication', {}),
('10378696:publication', 'publication', {}),
('15083194:publication', 'publication', {}),
('19029958:publication', 'publication', {}),
('22464321:publication', 'publication', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('22:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('94:patient', 'patient', {}),
('15:variant', 'variant', {'timestamp': '2023-06-21 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'FGFR1 amplification results in overexpression of the protein. This mutation has been found in squamous cell lung, bladder and breast cancer (PMID: 22960745, 26451490, 24476821). In vitro studies demonstrate that FGFR1 overexpression in breast cancer cell lines promotes resistance to the ESR1 inhibitor fulvestrant as measured by sustained cell viability upon drug treatment (PMID: 31371343). A patient with FGFR1-amplified small cell lung cancer had a partial response to the FGFR-inhibitor pazopanib (Abstract: Russo et al. JCO PO, 2019. https://ascopubs.org/doi/full/10.1200/PO.19.00243).', 'variant_summary': 'FGFR1 amplification is known to be oncogenic.'}),
('2:patient', 'patient', {}),
('FGFR1:gene_hugo', 'gene_hugo', {'entrez_gene_id': '2260', 'gene_summary': 'FGFR1, a receptor tyrosine kinase, is altered by mutation, chromosomal rearrangement or amplification in various cancer types including lung and breast cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'While there is promising clinical data supporting the use of pan-FGFR-targeted inhibitors such as erdafitinib in patients with high-level FGFR1-amplified squamous cell lung cancer, their clinical utility in patients with FGFR1-amplified high-grade serous ovarian cancer is unknown.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('31371343:publication', 'publication', {}),
('22960745:publication', 'publication', {}),
('24476821:publication', 'publication', {}),
('26451490:publication', 'publication', {}),
('Erdafitinib:drug', 'drug', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('30:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('2:patient', 'patient', {}),
('LEVEL_Fda3:fda_evidence_level', 'fda_evidence_level', {}),
('LEVEL_3B:oncokb_evidence_level', 'oncokb_evidence_level', {}),
('16:variant', 'variant', {'timestamp': '2019-08-28 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'Amplification of MYC is correlated with an increase in the expression of the MYC gene (PMID: 22464321,15083194), which leads to the transcriptional activation of MYC target genes. MYC target genes regulate various cellular processes involved in oncogenesis including proliferation, inhibiting exit from the cell cycle, stimulating vascularization and enhancing genomic instability (PMID: 22464321,10378696,16934487,19029958). Genomic amplification of MYC is observed in a large number of tumor types. A study of genomic amplification across many cancer types identified focal amplifications of MYC in 14% of all cancers, with additional enrichment for large scale arm-level copy number alterations encompassing the MYC gene locus (PMID: 20164920).', 'variant_summary': 'MYC amplification is known to be oncogenic.'}),
('41:patient', 'patient', {}),
('MYC:gene_hugo', 'gene_hugo', {'entrez_gene_id': '4609', 'gene_summary': 'MYC, a transcription factor, is altered by chromosomal rearrangement, amplification and overexpression in a variety of cancer types.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with MYC-amplified high-grade serous ovarian cancer.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('20164920:publication', 'publication', {}),
('16934487:publication', 'publication', {}),
('10378696:publication', 'publication', {}),
('15083194:publication', 'publication', {}),
('19029958:publication', 'publication', {}),
('22464321:publication', 'publication', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('85:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('41:patient', 'patient', {}),
('17:variant', 'variant', {'timestamp': '2021-10-05 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'Chromosome 9p24.1 has been shown to include JAK2 as well as the programmed cell death-1 (PD1) ligand genes, PD-L1 and PD-L2; its amplification is often detected in Classical Hodgkin lymphoma (cHL) and mediastinal large B-cell lymphoma (MLBCL). Amplification of 9p24.1 was associated with increased protein expression and activity as well as increased PD-1 ligand transcription and enhanced sensitivity to JAK2 inhibition (PMID: 20628145).', 'variant_summary': 'JAK2 amplification is known to be oncogenic.'}),
('40:patient', 'patient', {}),
('JAK2:gene_hugo', 'gene_hugo', {'entrez_gene_id': '3717', 'gene_summary': 'JAK2, an intracellular kinase, is frequently altered by mutation or chromosomal rearrangement in hematologic malignancies.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with JAK2-amplified high-grade serous ovarian cancer.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('20628145:publication', 'publication', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('57:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('40:patient', 'patient', {}),
('18:variant', 'variant', {'timestamp': '2019-01-16 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'Amplification of CD274 (more commonly PD-L1) has been detected across several tumor types including Hodgkin`s lymphoma (HL), mediastinal large B-cell lymphoma (MLBCL), small cell lung cancer, among others (PMID: 20628145, 29902298, 29377256, 27069084, 27620277). In preclinical assays, cells overexpressing PD-L1 exhibit increased features of immune evasion, due to binding of the PD-1 cell surface receptor expressed on immune cells (PMID: 28302645). In addition, the tyrosine kinase JAK2 is included in the 9p24.1 amplification region of PD-L1, which can also activate transcription of PD-L1 (PMID: 20628145). Preclinical studies have demonstrated that HL cells with PD-L1 amplification are sensitive to JAK2 inhibition (PMID: 24610827, 29467301). Patients with PD-L1 amplification have been found to have objective responses to PD-1/PD-L1-targeted immune therapies (PMID: 29902298, 28356247).', 'variant_summary': 'CD274 amplification is likely oncogenic.'}),
('48:patient', 'patient', {}),
('CD274:gene_hugo', 'gene_hugo', {'entrez_gene_id': '29126', 'gene_summary': 'CD274 (also known as PD-L1) is an immune receptor ligand. Expression of PD-L1 may help predict response to immunotherapies targeted against PD-L1 and its receptor, PD-1.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with CD274-amplified high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('28302645:publication', 'publication', {}),
('29377256:publication', 'publication', {}),
('20628145:publication', 'publication', {}),
('29902298:publication', 'publication', {}),
('27069084:publication', 'publication', {}),
('28356247:publication', 'publication', {}),
('24610827:publication', 'publication', {}),
('29467301:publication', 'publication', {}),
('27620277:publication', 'publication', {}),
('Likely Gain-of-function:functional_effect', 'functional_effect', {}),
('52:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('48:patient', 'patient', {}),
('19:variant', 'variant', {'timestamp': '2023-01-11 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'Amplification of PDCD1LG2 has been detected across several tumor types including Hodgkin`s lymphoma and mediastinal large B-cell lymphoma (PMID: 27069084, 29377256, 26850007, 26317899). In biochemical assays, cells overexpressing PDCD1LG2 demonstrate increased features of immune evasion from binding of the PD-1 cell surface receptor expressed on immune cells (PMID: 29167175). In addition, the tyrosine kinase JAK2 and PDCD1LG1 are included in the 9p24.1 amplification region of PDCD1LG2 (PMID: 20628145). Preclinical studies have demonstrated that cHL and MLBCL cell lines with PDCD1LG2 amplification are sensitive to JAK2 inhibition as measured by decreased tumor growth and prolonged survival (PMID: 24610827, 29467301). Patients with PDCD1LG2 amplification have been found to have an objective response to PD-1/PD-L1-targeted immune therapies (PMID: 29902298, 28356247).', 'variant_summary': 'PDCD1LG2 amplification is likely oncogenic.'}),
('31:patient', 'patient', {}),
('PDCD1LG2:gene_hugo', 'gene_hugo', {'entrez_gene_id': '80380', 'gene_summary': 'PDCD1LG2, a ligand of T-cell receptors involved in immune suppression, is overexpressed in various cancer types.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with PDCD1LG2-amplified high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('29377256:publication', 'publication', {}),
('20628145:publication', 'publication', {}),
('29902298:publication', 'publication', {}),
('27069084:publication', 'publication', {}),
('26850007:publication', 'publication', {}),
('28356247:publication', 'publication', {}),
('24610827:publication', 'publication', {}),
('29467301:publication', 'publication', {}),
('29167175:publication', 'publication', {}),
('26317899:publication', 'publication', {}),
('Likely Gain-of-function:functional_effect', 'functional_effect', {}),
('31:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('31:patient', 'patient', {}),
('20:variant', 'variant', {'timestamp': '2019-05-31 00:00:00', 'version': 'v4.10', 'variant_summary': 'CDKN2A amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('40:patient', 'patient', {}),
('CDKN2A:gene_hugo', 'gene_hugo', {'entrez_gene_id': '1029', 'gene_summary': 'The CDKN2A gene encodes two proteins, p16INK4A and p14ARF, that regulate the cell growth and survival. CDKN2A is altered by mutation and/or deletion in a broad range of solid and hematologic cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with CDKN2A-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('12:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('40:patient', 'patient', {}),
('21:variant', 'variant', {'timestamp': '2019-04-18 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'PIK3CA amplification is an infrequent event across cancers (PMID: 28481359), although several individual studies have reported them in lung squamous cell carcinoma, ovarian, esophagus, cervical, and head and neck tumors (PMID: 22960745, 21720365, 25631445, 23636398). PIK3CA gene amplification has been correlated with PIK3CA protein overexpression (PMID: 17992665,19697359). PIK3CA overexpression enhances growth and migration (PMID: 17803655, 26876212). PIK3CA amplifications associate with poor prognosis in several cancers (PMID: 22994622, 22292935, 27095573). Some studies have found a correlation between PIK3CA copy number amplification and sensitivity to PI3K inhibitors (PMID: 24608574, 23136191). Amplification of PIK3CA known pathogenic mutant alleles have also been reported (PMID: 19706770). Expression of this mutation in a patient-derived xenograft model of lung squamous cell carcinoma demonstrated that it was not sensitive to the pan-PIK3CA inhibitor BKM120 similar to PDX models expressing wildtype PIK3CA (PMID: 30093452).', 'variant_summary': 'PIK3CA amplification is likely oncogenic.'}),
('96:patient', 'patient', {}),
('PIK3CA:gene_hugo', 'gene_hugo', {'entrez_gene_id': '5290', 'gene_summary': 'PIK3CA, the catalytic subunit of PI3-kinase, is frequently mutated in a diverse range of cancers including breast, endometrial and cervical cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with PIK3CA-amplified high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('21720365:publication', 'publication', {}),
('22960745:publication', 'publication', {}),
('17992665:publication', 'publication', {}),
('23136191:publication', 'publication', {}),
('30093452:publication', 'publication', {}),
('19706770:publication', 'publication', {}),
('28481359:publication', 'publication', {}),
('22292935:publication', 'publication', {}),
('22994622:publication', 'publication', {}),
('19697359:publication', 'publication', {}),
('25631445:publication', 'publication', {}),
('24608574:publication', 'publication', {}),
('26876212:publication', 'publication', {}),
('23636398:publication', 'publication', {}),
('27095573:publication', 'publication', {}),
('17803655:publication', 'publication', {}),
('Likely Gain-of-function:functional_effect', 'functional_effect', {}),
('84:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('96:patient', 'patient', {}),
('22:variant', 'variant', {'timestamp': '2019-02-06 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'The amplification of BCL6 has been observed across a range of tumor types. Specifically, in follicular lymphoma BCL6 amplification was observed in a cohort of tumors lacking otherwise typical IGH/BCL2 rearrangements (PMID: 18500267, 24887414). BCL6 is a transcriptional regulator in germinal centers, from which human lymphomas often derive. In model systems, overexpression of BCL6 results in an anti-apoptotic and chemotherapy resistant phenotype (PMID: 15577913, 16142238).', 'variant_summary': 'BCL6 amplification is likely oncogenic.'}),
('55:patient', 'patient', {}),
('BCL6:gene_hugo', 'gene_hugo', {'entrez_gene_id': '604', 'gene_summary': 'BCL6, a transcriptional repressor involved in immune cell development, is frequently altered by chromosomal rearrangement in lymphomas.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with BCL6-amplified high-grade serous ovarian cancer.'}),
('Likely Oncogenic:oncogenicity', 'oncogenicity', {}),
('16142238:publication', 'publication', {}),
('18500267:publication', 'publication', {}),
('24887414:publication', 'publication', {}),
('15577913:publication', 'publication', {}),
('Likely Gain-of-function:functional_effect', 'functional_effect', {}),
('35:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('55:patient', 'patient', {}),
('23:variant', 'variant', {'timestamp': '2019-02-06 00:00:00', 'version': 'v4.10', 'variant_summary': 'TP63 amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('82:patient', 'patient', {}),
('TP63:gene_hugo', 'gene_hugo', {'entrez_gene_id': '8626', 'gene_summary': 'TP63, a transcription factor, is recurrently altered by chromosomal rearrangement in hematologic malignancies.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with TP63-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('23:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('82:patient', 'patient', {}),
('24:variant', 'variant', {'timestamp': '2019-07-30 00:00:00', 'version': 'v4.10', 'variant_summary': 'IL7R amplification has not specifically been reviewed by the OncoKB team, and therefore its biological significance is unknown.'}),
('54:patient', 'patient', {}),
('IL7R:gene_hugo', 'gene_hugo', {'entrez_gene_id': '3575', 'gene_summary': 'IL7R, a subunit of the interleukin 7 receptor, is altered in various hematologic malignancies including pediatric acute lymphoblastic leukemia.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'There are no FDA-approved or NCCN-compendium listed treatments specifically for patients with IL7R-amplified high-grade serous ovarian cancer.'}),
('Unknown:oncogenicity', 'oncogenicity', {}),
('Unknown:functional_effect', 'functional_effect', {}),
('67:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('54:patient', 'patient', {}),
('25:variant', 'variant', {'timestamp': '2023-06-21 00:00:00', 'version': 'v4.10', 'mutation_effect_description': 'FGFR1 amplification results in overexpression of the protein. This mutation has been found in squamous cell lung, bladder and breast cancer (PMID: 22960745, 26451490, 24476821). In vitro studies demonstrate that FGFR1 overexpression in breast cancer cell lines promotes resistance to the ESR1 inhibitor fulvestrant as measured by sustained cell viability upon drug treatment (PMID: 31371343). A patient with FGFR1-amplified small cell lung cancer had a partial response to the FGFR-inhibitor pazopanib (Abstract: Russo et al. JCO PO, 2019. https://ascopubs.org/doi/full/10.1200/PO.19.00243).', 'variant_summary': 'FGFR1 amplification is known to be oncogenic.'}),
('16:patient', 'patient', {}),
('FGFR1:gene_hugo', 'gene_hugo', {'entrez_gene_id': '2260', 'gene_summary': 'FGFR1, a receptor tyrosine kinase, is altered by mutation, chromosomal rearrangement or amplification in various cancer types including lung and breast cancers.'}),
('HGSOC:disease', 'disease', {'tumor_type_summary': 'While there is promising clinical data supporting the use of pan-FGFR-targeted inhibitors such as erdafitinib in patients with high-level FGFR1-amplified squamous cell lung cancer, their clinical utility in patients with FGFR1-amplified high-grade serous ovarian cancer is unknown.'}),
('Oncogenic:oncogenicity', 'oncogenicity', {}),
('31371343:publication', 'publication', {}),
('22960745:publication', 'publication', {}),
('24476821:publication', 'publication', {}),
('26451490:publication', 'publication', {}),
('Erdafitinib:drug', 'drug', {}),
('Gain-of-function:functional_effect', 'functional_effect', {}),
('68:sample', 'sample', {}),
('Amplification:alteration', 'alteration', {}),
('16:patient', 'patient', {}),
('LEVEL_Fda3:fda_evidence_level', 'fda_evidence_level', {}),
('LEVEL_3B:oncokb_evidence_level', 'oncokb_evidence_level', {}),
                    ]

    assert_edges = [('', '11:variant', '11753428:publication', 'published', {}),
                    ('', '0:variant', 'Capmatinib:drug', 'variant_affected_by_drug', {}),
                    ('', '20:variant', '12:sample', 'variant_in_sample', {}),
                    ('', '23:variant', '82:patient', 'patient_has_variant', {}),
                    ('', '7:variant', 'NTRK3:gene_hugo', 'variant_in_gene', {}),
                    ('', '14:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '4:variant', 'ETV6:gene_hugo', 'variant_in_gene', {}),
                    ('', '2:variant', 'C:oncogenicity', 'variant_to_oncogenicity', {'whatever': 'C2'}),
                    ('', '0:variant', '30073261:publication', 'published', {}),
                    ('', '6:variant', '29808010:publication', 'published', {}),
                    ('', '18:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '15:variant', '24476821:publication', 'published', {}),
                    ('', '10:variant', 'Y1463*:alteration', 'variant_to_alteration', {}),
                    ('', '1:variant', '24055054:publication', 'published', {}),
                    ('', '25:variant', 'LEVEL_Fda3:fda_evidence_level', 'variant_to_fda_evidence', {}),
                    ('', '13:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '13:variant', '26451490:publication', 'published', {}),
                    ('', '15:variant', '30:sample', 'variant_in_sample', {}),
                    ('', '14:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '1:variant', 'B:patient', 'patient_has_variant', {}),
                    ('', '25:variant', '24476821:publication', 'published', {}),
                    ('', '12:variant', '22012066:publication', 'published', {}),
                    ('', '1:variant', 'BRAF:gene_hugo', 'variant_in_gene', {}),
                    ('', '10:variant', '12483515:publication', 'published', {}),
                    ('', '15:variant', 'LEVEL_3B:oncokb_evidence_level', 'variant_to_oncokb_evidence', {}),
                    ('', '12:variant', '52:sample', 'variant_in_sample', {}),
                    ('', '19:variant', '27069084:publication', 'published', {}),
                    ('', '22:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '18:variant', '24610827:publication', 'published', {}),
                    ('', '1:variant', '21098728:publication', 'published', {}),
                    ('', '92:sample', '88:patient', 'sample_to_patient', {}),
                    ('', '22:variant', '16142238:publication', 'published', {}),
                    ('', '14:variant', '22464321:publication', 'published', {}),
                    ('', '21:variant', '22292935:publication', 'published', {}),
                    ('', '8:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '25:variant', '68:sample', 'variant_in_sample', {}),
                    ('', '15:variant', 'LEVEL_Fda3:fda_evidence_level', 'variant_to_fda_evidence', {}),
                    ('', '16:variant', '85:sample', 'variant_in_sample', {}),
                    ('', '18:variant', '28356247:publication', 'published', {}),
                    ('', '19:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '0:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '2:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '3:variant', '60:patient', 'patient_has_variant', {}),
                    ('', '17:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '12:variant', 'BCOR:gene_hugo', 'variant_in_gene', {}),
                    ('', '20:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '18:variant', '52:sample', 'variant_in_sample', {}),
                    ('', '19:variant', '29467301:publication', 'published', {}),
                    ('', '7:variant', '65:patient', 'patient_has_variant', {}),
                    ('', '3:variant', 'Loss-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '11:variant', 'Likely Loss-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '18:variant', '29902298:publication', 'published', {}),
                    ('', '24:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '16:variant', 'MYC:gene_hugo', 'variant_in_gene', {}),
                    ('', '9:variant', 'MAF:gene_hugo', 'variant_in_gene', {}),
                    ('', '6:variant', '30072474:publication', 'published', {}),
                    ('', '25:variant', '26451490:publication', 'published', {}),
                    ('', '37:sample', '60:patient', 'sample_to_patient', {}),
                    ('', '8:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '10:variant', 'Likely Loss-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '8:variant', '13:sample', 'variant_in_sample', {}),
                    ('', '7:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '21:variant', '84:sample', 'variant_in_sample', {}),
                    ('', '21:variant', '23136191:publication', 'published', {}),
                    ('', '9:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '16:variant', '15083194:publication', 'published', {}),
                    ('', '15:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '16:variant', '10378696:publication', 'published', {}),
                    ('', '22:variant', '18500267:publication', 'published', {}),
                    ('', '18:variant', '20628145:publication', 'published', {}),
                    ('', '10:variant', 'BRCA1:gene_hugo', 'variant_in_gene', {}),
                    ('', '25:variant', '31371343:publication', 'published', {}),
                    ('', '16:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '18:variant', '48:patient', 'patient_has_variant', {}),
                    ('', '22:sample', '94:patient', 'sample_to_patient', {}),
                    ('', '25:variant', 'FGFR1:gene_hugo', 'variant_in_gene', {}),
                    ('', '6:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '14:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '16:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '8:variant', '35:patient', 'patient_has_variant', {}),
                    ('', '24:variant', '54:patient', 'patient_has_variant', {}),
                    ('', '1:variant', '67:patient', 'patient_has_variant', {}),
                    ('', '5:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '9:variant', '75:patient', 'patient_has_variant', {}),
                    ('', '3:variant', 'Ribociclib:drug', 'variant_affected_by_drug', {}),
                    ('', '10:variant', 'Olaparib:drug', 'variant_affected_by_drug', {}),
                    ('', '11:variant', '11900253:publication', 'published', {}),
                    ('', '12:variant', '24047651:publication', 'published', {}),
                    ('', '0:variant', '19117057:publication', 'published', {}),
                    ('', '13:variant', '22960745:publication', 'published', {}),
                    ('', '4:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '15:variant', '26451490:publication', 'published', {}),
                    ('', '0:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '24:variant', 'IL7R:gene_hugo', 'variant_in_gene', {}),
                    ('', '22:variant', '24887414:publication', 'published', {}),
                    ('', '11:variant', '16007150:publication', 'published', {}),
                    ('', '4:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '3:variant', 'Abemaciclib:drug', 'variant_affected_by_drug', {}),
                    ('', '6:variant', '23099803:publication', 'published', {}),
                    ('', '14:variant', '16934487:publication', 'published', {}),
                    ('', '13:variant', '1:sample', 'variant_in_sample', {}),
                    ('', '21:variant', '28481359:publication', 'published', {}),
                    ('', '11:variant', 'TP53:gene_hugo', 'variant_in_gene', {}),
                    ('', '18:variant', '28302645:publication', 'published', {}),
                    ('', '12:variant', '25550361:publication', 'published', {}),
                    ('', '13:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '46:sample', '65:patient', 'sample_to_patient', {}),
                    ('', '52:sample', '77:patient', 'sample_to_patient', {}),
                    ('', '23:sample', '82:patient', 'sample_to_patient', {}),
                    ('', '22:variant', '55:patient', 'patient_has_variant', {}),
                    ('', '12:variant', '26847029:publication', 'published', {}),
                    ('', '25:variant', 'LEVEL_3B:oncokb_evidence_level', 'variant_to_oncokb_evidence', {}),
                    ('', '17:variant', '20628145:publication', 'published', {}),
                    ('', '2:variant', '88:patient', 'patient_has_variant', {}),
                    ('', '5:variant', 'ETNK1:gene_hugo', 'variant_in_gene', {}),
                    ('', '1:variant', '22395615:publication', 'published', {}),
                    ('', '11:variant', '1:sample', 'variant_in_sample', {}),
                    ('', '19:variant', '29902298:publication', 'published', {}),
                    ('', '14:variant', '20164920:publication', 'published', {}),
                    ('', '9:variant', '85:sample', 'variant_in_sample', {}),
                    ('', '0:variant', '28481359:publication', 'published', {}),
                    ('', '11:variant', '21467160:publication', 'published', {}),
                    ('', '2:variant', '21837672:publication', 'published', {}),
                    ('', '43:sample', '40:patient', 'sample_to_patient', {}),
                    ('', '18:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '21:variant', '17992665:publication', 'published', {}),
                    ('', '17:variant', '40:patient', 'patient_has_variant', {}),
                    ('', '0:variant', '18077425:publication', 'published', {}),
                    ('', '0:variant', '18093943:publication', 'published', {}),
                    ('', '21:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '2:variant', '92:sample', 'variant_in_sample', {}),
                    ('', '7:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '19:variant', '20628145:publication', 'published', {}),
                    ('', '1:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '5:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '0:variant', 'Telisotuzumab Vedotin:drug', 'variant_affected_by_drug', {}),
                    ('', '0:variant', '17463250:publication', 'published', {}),
                    ('', '19:variant', '29377256:publication', 'published', {}),
                    ('', '1:sample', '11:patient', 'sample_to_patient', {}),
                    ('', '19:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '23:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '7:variant', '46:sample', 'variant_in_sample', {}),
                    ('', '21:variant', '26876212:publication', 'published', {}),
                    ('', '1:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '31:sample', '31:patient', 'sample_to_patient', {}),
                    ('', '3:variant', '21526190:publication', 'published', {}),
                    ('', '6:variant', '68:patient', 'patient_has_variant', {}),
                    ('', '1:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '13:variant', '31371343:publication', 'published', {}),
                    ('', '23:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '14:variant', '22:sample', 'variant_in_sample', {}),
                    ('', '16:variant', '22464321:publication', 'published', {}),
                    ('', '20:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '4:sample', '67:patient', 'sample_to_patient', {}),
                    ('', '20:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '14:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '14:variant', '10378696:publication', 'published', {}),
                    ('', '2:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '5:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '22:variant', 'Likely Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '2:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '10:variant', 'Talazoparib:drug', 'variant_affected_by_drug', {}),
                    ('', '6:variant', '23404247:publication', 'published', {}),
                    ('', '15:variant', '2:patient', 'patient_has_variant', {}),
                    ('', '14:variant', '15083194:publication', 'published', {}),
                    ('', '22:variant', 'BCL6:gene_hugo', 'variant_in_gene', {}),
                    ('', '76:sample', '68:patient', 'sample_to_patient', {}),
                    ('', '25:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '4:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '10:variant', '61:patient', 'patient_has_variant', {}),
                    ('', '6:variant', 'KRAS:gene_hugo', 'variant_in_gene', {}),
                    ('', '19:variant', '26317899:publication', 'published', {}),
                    ('', '14:variant', '94:patient', 'patient_has_variant', {}),
                    ('', '4:variant', '15:sample', 'variant_in_sample', {}),
                    ('', '13:variant', '5:patient', 'patient_has_variant', {}),
                    ('', '1:variant', '4:sample', 'variant_in_sample', {}),
                    ('', '17:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '19:variant', '24610827:publication', 'published', {}),
                    ('', '21:variant', '96:patient', 'patient_has_variant', {}),
                    ('', '16:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '22:variant', '35:sample', 'variant_in_sample', {}),
                    ('', '15:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '23:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '2:variant', 'EZH2:gene_hugo', 'variant_in_gene', {}),
                    ('', '22:variant', '15577913:publication', 'published', {}),
                    ('', '13:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '21:variant', 'Likely Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '9:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '12:sample', '40:patient', 'sample_to_patient', {}),
                    ('', '7:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '18:variant', '29377256:publication', 'published', {}),
                    ('', '19:variant', '31:patient', 'patient_has_variant', {}),
                    ('', '10:variant', '12947386:publication', 'published', {}),
                    ('', '8:sample', '61:patient', 'sample_to_patient', {}),
                    ('', '21:variant', '30093452:publication', 'published', {}),
                    ('', '14:variant', 'MYC:gene_hugo', 'variant_in_gene', {}),
                    ('', '12:variant', '77:patient', 'patient_has_variant', {}),
                    ('', '3:variant', 'CDKN2A:gene_hugo', 'variant_in_gene', {}),
                    ('', '0:variant', 'Crizotinib:drug', 'variant_affected_by_drug', {}),
                    ('', '11:variant', '27759562:publication', 'published', {}),
                    ('', '25:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '10:variant', 'Rucaparib:drug', 'variant_affected_by_drug', {}),
                    ('', '3:variant', 'LEVEL_Fda3:fda_evidence_level', 'variant_to_fda_evidence', {}),
                    ('', '5:variant', '43:sample', 'variant_in_sample', {}),
                    ('', '11:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '6:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '15:variant', '31371343:publication', 'published', {}),
                    ('', '5:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '19:variant', '26850007:publication', 'published', {}),
                    ('', '19:variant', 'Likely Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '19:variant', '29167175:publication', 'published', {}),
                    ('', '4:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '22:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '0:variant', 'LEVEL_3B:oncokb_evidence_level', 'variant_to_oncokb_evidence', {}),
                    ('', '16:variant', '19029958:publication', 'published', {}),
                    ('', '0:variant', '22869872:publication', 'published', {}),
                    ('', '21:variant', '25631445:publication', 'published', {}),
                    ('', '0:variant', 'A:disease', 'variant_to_disease', {'something': 'Whatever it is'}),
                    ('', '6:variant', '76:sample', 'variant_in_sample', {}),
                    ('', '0:variant', 'Tepotinib:drug', 'variant_affected_by_drug', {}),
                    ('', '17:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '84:sample', '96:patient', 'sample_to_patient', {}),
                    ('', '11:variant', '19336573:publication', 'published', {}),
                    ('', '2:variant', '16330673:publication', 'published', {}),
                    ('', '8:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '12:variant', 'Likely Loss-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '18:variant', '29467301:publication', 'published', {}),
                    ('', '18:variant', '27069084:publication', 'published', {}),
                    ('', '21:variant', '27095573:publication', 'published', {}),
                    ('', '12:variant', 'S1704Kfs*16:alteration', 'variant_to_alteration', {}),
                    ('', '15:variant', 'FGFR1:gene_hugo', 'variant_in_gene', {}),
                    ('', '7:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '20:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '21:variant', '22960745:publication', 'published', {}),
                    ('', '15:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '15:variant', '22960745:publication', 'published', {}),
                    ('', '0:variant', '16461907:publication', 'published', {}),
                    ('', '22:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '85:sample', '75:patient', 'sample_to_patient', {}),
                    ('', '19:variant', '31:sample', 'variant_in_sample', {}),
                    ('', '15:variant', 'Erdafitinib:drug', 'variant_affected_by_drug', {}),
                    ('', '0:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '8:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '10:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '8:variant', 'IDH2:gene_hugo', 'variant_in_gene', {}),
                    ('', '10:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '3:variant', '37:sample', 'variant_in_sample', {}),
                    ('', '16:variant', '20164920:publication', 'published', {}),
                    ('', '91:sample', '33:patient', 'sample_to_patient', {}),
                    ('', '0:variant', 'MET:gene_hugo', 'variant_in_gene', {}),
                    ('', '11:variant', 'X125_splice:alteration', 'variant_to_alteration', {}),
                    ('', '30:sample', '2:patient', 'sample_to_patient', {}),
                    ('', '16:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '2:variant', '23529930:publication', 'published', {}),
                    ('', '13:variant', 'Erdafitinib:drug', 'variant_affected_by_drug', {}),
                    ('', '12:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '0:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '18:variant', '27620277:publication', 'published', {}),
                    ('', '25:variant', 'Erdafitinib:drug', 'variant_affected_by_drug', {}),
                    ('', '10:variant', 'Niraparib:drug', 'variant_affected_by_drug', {}),
                    ('', '20:variant', '40:patient', 'patient_has_variant', {}),
                    ('', '23:variant', 'TP63:gene_hugo', 'variant_in_gene', {}),
                    ('', '25:variant', 'Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '21:variant', '19697359:publication', 'published', {}),
                    ('', '23:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '1:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '1:sample', '5:patient', 'sample_to_patient', {}),
                    ('', '6:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '85:sample', '41:patient', 'sample_to_patient', {}),
                    ('', '3:variant', '26516359:publication', 'published', {}),
                    ('', '0:variant', '91:sample', 'variant_in_sample', {}),
                    ('', '16:variant', '16934487:publication', 'published', {}),
                    ('', '3:variant', 'LEVEL_4:oncokb_evidence_level', 'variant_to_oncokb_evidence', {}),
                    ('', '12:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '13:variant', 'LEVEL_3B:oncokb_evidence_level', 'variant_to_oncokb_evidence', {}),
                    ('', '10:variant', '20608970:publication', 'published', {}),
                    ('', '0:variant', 'LEVEL_Fda3:fda_evidence_level', 'variant_to_fda_evidence', {}),
                    ('', '3:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '24:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '13:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '10:variant', 'LEVEL_Fda2:fda_evidence_level', 'variant_to_fda_evidence', {}),
                    ('', '68:sample', '16:patient', 'sample_to_patient', {}),
                    ('', '18:variant', 'Likely Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '21:variant', '24608574:publication', 'published', {}),
                    ('', '24:variant', 'Unknown:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '24:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '14:variant', '19029958:publication', 'published', {}),
                    ('', '3:variant', 'Deletion:alteration', 'variant_to_alteration', {}),
                    ('', '10:variant', '11358863:publication', 'published', {}),
                    ('', '2:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '25:variant', '16:patient', 'patient_has_variant', {}),
                    ('', '10:variant', 'LEVEL_1:oncokb_evidence_level', 'variant_to_oncokb_evidence', {}),
                    ('', '1:variant', '15467732:publication', 'published', {}),
                    ('', '13:variant', 'LEVEL_Fda3:fda_evidence_level', 'variant_to_fda_evidence', {}),
                    ('', '3:variant', '28982163:publication', 'published', {}),
                    ('', '15:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '4:variant', '32:patient', 'patient_has_variant', {}),
                    ('', '20:variant', 'CDKN2A:gene_hugo', 'variant_in_gene', {}),
                    ('', '3:variant', '10999737:publication', 'published', {}),
                    ('', '21:variant', '23636398:publication', 'published', {}),
                    ('', '21:variant', '19706770:publication', 'published', {}),
                    ('', '13:variant', 'FGFR1:gene_hugo', 'variant_in_gene', {}),
                    ('', '6:variant', 'Likely Gain-of-function:functional_effect', 'variant_has_effect', {}),
                    ('', '19:variant', '28356247:publication', 'published', {}),
                    ('', '6:variant', '20841479:publication', 'published', {}),
                    ('', '21:variant', 'PIK3CA:gene_hugo', 'variant_in_gene', {}),
                    ('', '0:variant', '17667909:publication', 'published', {}),
                    ('', '21:variant', '17803655:publication', 'published', {}),
                    ('', '24:variant', '67:sample', 'variant_in_sample', {}),
                    ('', '21:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '21:variant', '22994622:publication', 'published', {}),
                    ('', '18:variant', 'CD274:gene_hugo', 'variant_in_gene', {}),
                    ('', '11:variant', '11:patient', 'patient_has_variant', {}),
                    ('', '5:variant', '40:patient', 'patient_has_variant', {}),
                    ('', '67:sample', '54:patient', 'sample_to_patient', {}),
                    ('', '10:variant', '8:sample', 'variant_in_sample', {}),
                    ('', '25:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '25:variant', '22960745:publication', 'published', {}),
                    ('', '11:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '2:variant', '14500907:publication', 'published', {}),
                    ('', '57:sample', '40:patient', 'sample_to_patient', {}),
                    ('', '0:variant', '33:patient', 'patient_has_variant', {}),
                    ('', '9:variant', 'Unknown:functional_effect', 'variant_has_effect', {}),
                    ('', '15:sample', '32:patient', 'sample_to_patient', {}),
                    ('', '23:variant', '23:sample', 'variant_in_sample', {}),
                    ('', '17:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '21:variant', '21720365:publication', 'published', {}),
                    ('', '3:variant', 'Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '18:variant', 'Amplification:alteration', 'variant_to_alteration', {}),
                    ('', '16:variant', '41:patient', 'patient_has_variant', {}),
                    ('', '13:sample', '35:patient', 'sample_to_patient', {}),
                    ('', '2:variant', '12374981:publication', 'published', {}),
                    ('', '17:variant', '57:sample', 'variant_in_sample', {}),
                    ('', '21:variant', 'Likely Oncogenic:oncogenicity', 'variant_to_oncogenicity', {}),
                    ('', '19:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '19:variant', 'PDCD1LG2:gene_hugo', 'variant_in_gene', {}),
                    ('', '1:variant', '25618114:publication', 'published', {}),
                    ('', '35:sample', '55:patient', 'sample_to_patient', {}),
                    ('', '17:variant', 'JAK2:gene_hugo', 'variant_in_gene', {}),
                    ('', '52:sample', '48:patient', 'sample_to_patient', {}),
                    ('', '3:variant', 'Palbociclib:drug', 'variant_affected_by_drug', {}),
                    ('', '12:variant', '22237022:publication', 'published', {}),
                    ('', '9:variant', 'HGSOC:disease', 'variant_to_disease', {}),
                    ('', '13:variant', '24476821:publication', 'published', {}),
                    ('', '3:variant', '15495191:publication', 'published', {}),
                    ]

    data_mapping = {f"tests/{directory_name2}/data.csv": f"tests/{directory_name2}/mapping.yaml",
                    f"tests/{directory_name1}/data.csv": f"tests/{directory_name1}/mapping.yaml"}

    nodes, edges = ontoweaver.extract(data_mapping, affix="suffix", validate_output=True, raise_errors=False)
    assert nodes, "There is no node"
    assert edges, "There is no edge"

    bc_nodes = [n.as_tuple() for n in nodes]
    assert_node_set = testing_functions.convert_to_set(assert_nodes)
    f_node_set = testing_functions.convert_to_set(bc_nodes)
    assert assert_node_set == f_node_set, "Nodes are not equal."

    bc_edges = [e.as_tuple() for e in edges]
    assert_edge_set = testing_functions.convert_to_set(assert_edges)
    f_edge_set = testing_functions.convert_to_set(bc_edges)
    assert assert_edge_set == f_edge_set, "Edges are not equal."


if __name__ == "__main__":
    test_2_databases()
